pageserver/src/tenant/timeline.rs

```diff
@@ -2637,6 +2639,7 @@ impl Timeline {
         let layers = self.layers.read().unwrap();
 
         let mut max_deltas = 0;
+        let wanted_image_layers = self.wanted_image_layers.lock().unwrap().take();
```

Comment:
Design-wise this is a bit problematic that a query method (returning bool) modifies the state so running the query twice would have different results, but that could be moved upwards, and probably should be moved upwards for the plural of locks.

Reply:
I agree with you that functions with side effects are more obscure and difficult for understanding and verification than pure functions without side effects. But please notice that this function is not idempotent i any case, because it depends on current layer map state.

Sorry, but I do not know better solution. We should somehow pass information about required image layers from GC to compaction task. It needs to be done only once, because if required image layers are not generated at this iteration of compaction, then there are no reasons to generate them at subsequent iterations. So there is no sense to try to keep this map and perform this checks at all compaction iteration until next GC (which is expected to be done less frequently than compaction because of larger PiTR interval).

So, the alternative to `take`  is to iterate through layers under `wanted_image_layers` lock.
But it is always better to hold a lock as short as possible.
And as I wrote above there are good reasons to perform this check only once, taken in account that number of compaction iterations is expected to be several times larger than GC iterations.